### DISCUSSION TITLE
Socket error by SQLAlchemy but not by PyMySQL

### DISCUSSION THREAD

**Discussion by hy144328** [sqlalchemy/sqlalchemy#7777] [Usage Questions] [answered]
I am trying to run SQLAlchemy inside a Docker container on my company's Kubernetes environment. This is the relevant section of my script:

```
engine = sqla.create_engine(
    "mysql+pymysql://{}:{}@{}:{}/{}?charset=utf8mb4".format(
        os.getenv("MYSQL_USER"),
        os.getenv("MYSQL_PASS"),
        os.getenv("MYSQL_HOST"),
        os.getenv("MYSQL_PORT"),
        os.getenv("MYSQL_DB"),
    ),
)
with engine.connect() as conn:
    # conn.execute(q)
```

But I get an error message (see bottom).

Nevertheless, when I get the connection straight from PyMySQL, I do not run into any issues:

```
connection = pymysql.connect(
    host=os.getenv("MYSQL_HOST"),
    user=os.getenv("MYSQL_USER"),
    password=os.getenv("MYSQL_PASS"),
    database=os.getenv("MYSQL_DB"),
    charset='utf8mb4',
    cursorclass=pymysql.cursors.DictCursor,
)    
with connection:
    with connection.cursor() as cursor:
        # cursor.execute("SELECT COUNT(*) FROM ...")
```

Questions:

1.  Am I missing a setting in SQLAlchemy?
2.  Is there a bug in SQLAlchemy? Because PyMySQL seem to be working fine. I have tried other drivers from the list, and I get similar error messages?

I apologize for not providing a stand-alone example. This is something that happens (i) on my Minikube Kubernetes cluster, and (ii) on my company's Kubernetes cluster -- not on my desktop when I run it outside any container. Nevertheless, I do suspect that SQLAlchemy does something under the hood differently from PyMySQL. Otherwise, I have no explanation. Please let me know if there is anything else that I should provide. Thank you in advance for any input!

**Comment by CaselIT** [comment, marked as answer]
Hi,

This may be caused by special characters in the user or password. Try to url encode them like mentioned in the docs https://docs.sqlalchemy.org/en/14/core/engines.html#database-urls

**Reply by hy144328** [reply]
Essentially the right answer although slightly more nuanced:

I used a `.env` file where I put the value of each key-value pair in double quotes.
For development, works fine with `python-dotenv`.
For production, I use `kubectl create secret generic --from-env-file=.env` to turn the `.env` file into a secret.
Turns out that `kubectl`, unlike `python-dotenv`, does not drop the double quotes.
This messes with the connection string.

Lessons learnt:

*   Watch out for special characters in `os.environ` when reading environment variables (`python-dotenv`, `kubectl`) and writing connection strings (`urllib.parse`).
*   Drop quotes in `.env` files.

Hope this helps another programmer.
